Apply Least Privilege & RBAC for Ansible Execution + Security Enhancements

daryl375 opened the pull request:
# Ansible Security Improvements and RBAC Implementation

This PR introduces security improvements and a new RBAC-based execution model for Ansible playbooks. The goal is to:

- Apply the Principle of Least Privilege
- Strengthen control over who can execute what through Ansible

## Key Security Enhancements

- **Salt length increased:** from 64 bits to 256 bits
- **SHA256 iterations:** set to 1,000
- **Input validation:** simplified for user security questions and answers
- **Rate limiting:** users are allowed a maximum of 3 failed attempts
- **Race condition fix:** during secure directory creation

### Account Lockout Policy

If a user fails the security question 3 times, `sudo reset-my-password` will be disabled for 24 hours.

## Role-Based Access Control (RBAC) via ansible_executor Group

To enforce RBAC and apply least privilege, a new system group has been added:

- **Group:** `ansible_executor`
- **Sudoers file:** `/etc/sudoers.d/40-ansible-executor`
- **Allowed command:**

```bash
# /etc/sudoers.d/40-ansible-executor
%ansible_executor ALL=(ALL) PASSWD: /bin/sh
```

This grants users in the `ansible_executor` group the ability to run `/bin/sh` with sudo, but only if they have a valid password configured (set through the self-service password mechanism).

### Technical Justification

This permission is essential because Ansible does not execute individual system commands directly. Instead, it wraps tasks in a temporary Python script and runs that script using `/bin/sh` as the shell interpreter.

For example, a task like:

```yaml
- name: Test 5 - Test service control with ansible_executor
  ansible.builtin.service:
    name: cron
    state: started
  become: true
  register: service_test
  ignore_errors: yes
```

does not translate into a direct `systemctl start cron` command. Instead, Ansible internally executes something similar to:

```bash
sudo -H -S -n -u root /bin/sh -c 'echo BECOME-SUCCESS-... ; /usr/bin/python3.8 /home/<user>/.ansible/tmp/.../AnsiballZ_*.py'
```

If permission to run `/bin/sh` via sudo is not granted, all such module executions will fail, regardless of the task being run.

### Official Documentation Reference

According to the [Ansible documentation](https://docs.ansible.com/ansible/latest/playbook_guide/playbooks_privilege_escalation.html#privilege-escalation-must-be-general):

> **Privilege escalation must be general**
> 
> You cannot limit privilege escalation permissions to certain commands. Ansible does not always use a specific command to do something but runs modules (code) from a temporary file name which changes every time. If you have '/sbin/service' or '/bin/chmod' as the allowed commands this will fail with Ansible as those paths won't match with the temporary file that Ansible creates to run the module. If you have security rules that constrain your sudo/pbrun/doas environment to run specific command paths only, use Ansible from a special account that does not have this constraint, or use AWX or the Red Hat Ansible Automation Platform to manage indirect access to SSH credentials.

## Recommended Pre-check for Playbooks

To ensure that the executing user has a valid password, add this `pre_tasks` block to your playbooks:

```yaml
pre_tasks:
  - name: Get Ansible user groups
    ansible.builtin.shell: 'groups {{ ansible_user }}'
    register: ansible_user_groups
    run_once: true

  - name: Check if user belongs to ansible_executor group
    ansible.builtin.debug:
      msg: "User {{ ansible_user }} groups: {{ ansible_user_groups.stdout }}"
    run_once: true

  - name: Check password status
    ansible.builtin.shell: "passwd -S {{ ansible_user }} | awk '{print $2}'"
    register: password_status
    run_once: true

  - name: Debug password status
    ansible.builtin.debug:
      msg: "User {{ ansible_user }} password status: {{ password_status.stdout }}"
    run_once: true

  - name: Fail if ansible_executor user has no password
    ansible.builtin.fail:
      msg: |
        SECURITY WARNING: You are executing this playbook as user '{{ ansible_user }}' who belongs to the 'ansible_executor' group,
        but you have not set a password for your account. This is a security risk.

        To fix this, please:
        1. SSH into the server:
           ssh -p 2522 {{ ansible_user }}@{{ ansible_default_ipv4.address }}
        2. Run the self-service reset:
           sudo reset-my-password

        This check prevents execution with insufficient authentication.
    when: >
      "'ansible_executor' in ansible_user_groups.stdout" and
      ("NP" in password_status.stdout or "L" in password_status.stdout)
    run_once: true
```

## Usage Instructions

### Use -K When Running Playbooks

Always execute playbooks that rely on sudo with the `-K` flag to prompt for the user's password:

```bash
ansible-playbook playbooks/pb_test_ansible_executor.yml \
  -i solana_new_metal_box.yml \
  -e "target_host=new-metal-box" \
  -e "ansible_user=yamel" \
  -K
```

The `-K` flag is mandatory and will prompt you to enter your password before executing any privileged operations.

---

# Execute playbook to configure users using Local Cluster

## Prerequisites
- After initializing the localcluster

## Previous configuration on target server

Before executing the playbook from localcluster, you must first access the server you are going to configure (example: host-charlie).

1. Access with user `sol`
2. Execute `sudo visudo`
3. Edit the line:
   ```
   sol ALL=(ALL) NOPASSWD: ALL
   ```
   Replacing `sol` with the user that comes provisioned in the local cluster

## CSV example

```bash
cat ~/new-metal-box/users3.csv
```

```csv
user,key,group_a,group_b,group_c,group_d
john,ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIFzK8w9H2mTqjXzP4YvN7bQrS1cDgEhW6fR3uLpM9nXy,sysadmin,,,
sarah,ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIGpN5t7RqW3kJ8mY2sC6vFhX9zAcE1bQ4dLnP0oU7iMf,sysadmin,validator_operators,validator_admins,
mike,ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABgQDHq8K9mL3nP6tY2wR5jF4cG7eV1bX9sN0uM8pQ2iE,validator_operators,ansible_executor,,
lisa,ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIJ7eR4mN9qW6tK3sF1bY5cP2vL8hG0dX3nU7jZa1rMpI,sysadmin,,,
david,ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABgQC2mK8tR9jP4nY6wL3sF7eV5bX1cN2uM0pQ8iE9,validator_viewers,,,
```

## Testing considerations

- For testing purposes you can put your public key in all users to be able to access each one of them once the server is provisioned and do the necessary tests with each user

- The localcluster containers come provisioned with your user using an environment variable, so in the CSV you should substitute your name for another one since the playbook in one of its prechecks verifies that none of the users provided from the CSV are on the target host, preventing it from being executed on a server that is already configured

## Command execution

```bash
cd ansible/
ansible-playbook playbooks/pb_setup_server_users.yml \
  -i solana_localnet.yml \
  -e "target_host=host-charlie" \
  -e "ansible_user=sol" \
  -e "csv_file=users3.csv"
```

daryl375 commented:
Changes applied based on Copilot recommendations:
40-ansible-executor.j2
Removed misleading comment:
The comment "Allow ansible_executor to execute all commands without password" was removed, as the associated sudoers directive used PASSWD: (which does require a password). The comment was inconsistent with the configuration and could lead to confusion.

precheck.yml
Explicit CSV field validation:
Replaced dynamic map(attribute=item) logic with explicit checks for each required field (e.g., group_a, shell, uid, etc.).
This prevents unintended behavior where item is treated as a variable name instead of a string and ensures Ansible reliably detects missing fields.

docker-compose.yml
Use absolute path for volume mount:
Docker Compose does not expand ~ in volume paths.
Updated the volume definition from:

```yaml
- ~/new-metal-box:/root/new-metal-box
```
to:
```yaml
- ${HOME}/new-metal-box:/root/new-metal-box
This ensures that the user CSV directory is mounted correctly in all environments, including CI/CD pipelines or container orchestrators.
```

solana_new_metal_box.yml
No change needed for ansible_host:
Although Copilot flagged the ansible_host line as misaligned, this is a false positive.
The structure is valid — ansible_host is correctly nested under the new-metal-box host as per Ansible's static inventory syntax.
Verified using:
```bash
ansible-inventory -i solana_new_metal_box.yml --list
```

MexicanSparker commented on ansible/roles/iam_manager/templates/reset-my-password.j2:
```diff
@@ -0,0 +1,392 @@
+    chmod 700 "$SEC_DIR"
+fi
+
+# Check for user lockout (24-hour block or until admin intervention)
+BLOCK_FILE="/tmp/password-reset-blocked-$SUDO_USER"
```
The password reset block file is stored in `/tmp` by the line, which is world-writable and may be cleared on reboot. This could allow a user to remove their own block or bypass lockout by rebooting.

```bash
BLOCK_FILE="/tmp/password-reset-blocked-$SUDO_USER"
```

If you want to make the block persistent and more secure, you should change this path to something like `/var/lib/password-security/password-reset-blocked-$SUDO_USER` (and ensure the directory exists and is properly permissioned).

daryl375 replied:
Changed password reset block file location to /etc/password-security/

MexicanSparker commented on ansible/roles/iam_manager/tasks/create_users.yml:
```diff
@@ -1,55 +1,52 @@
-- name: Add users to groups
+# Create users from CSV data with RBAC groups
+- name: Create users from CSV
   ansible.builtin.user:
     name: "{{ item.key }}"
-    groups: "{{ (item.value.group_a | default('') | lower + ',' + item.value.group_b | default('') | lower) | regex_replace('^,|,$', '') | regex_replace(',,', ',') | regex_replace('^$', '') }}"
+    shell: /bin/bash
+    state: present
+    groups: "{{ (item.value.group_a | default('') + ',' + item.value.group_b | default('') + ',' + item.value.group_c | default('') + ',' + item.value.group_d | default('')) | regex_replace('^,|,$', '') | regex_replace(',,', ',') | regex_replace('^$', '') }}"
     append: true
   loop: "{{ users_csv.dict | dict2items }}"
-  when: item.value.create_user | default(true) and ((item.value.group_a | default('') != '') or (item.value.group_b | default('') != ''))
   no_log: true
```
## Potential for humans to introduce weak keys

There is no explicit check for key length or type beyond the prefix. Weak keys (e.g., short RSA) could be accepted if present in the CSV.

**How to Enforce Key Strength in Ansible**

You can add a task to check the key type and, for RSA, the length. For example, using a custom filter or a shell command:

```yaml
# This is a simplified example; a robust check would require a custom module or script.
- name: Fail if RSA key is too short
  ansible.builtin.fail:
    msg: "RSA key for user {{ item.key }} is too short (must be at least 2048 bits)"
  when: >
    item.value.key.startswith('ssh-rsa ') and
    (lookup('pipe', "echo '{{ item.value.key }}' | awk '{print $2}' | base64 -d | od -An -t u1 | awk '{print $2}'") | int) < 2048
```

eydelrivero reviewed (changes requested):
This PR needs to be updated now that the `host-charlie` node in the localnet cluster has been fixed to replicate an Ubuntu environment, similar to what is typically provided by bare-metal providers. Accordingly, the PR description should also be revised to reflect this change.
In its current form, reviewing this large PR requires numerous manual steps, particularly around setting up an IAM `users.csv` file containing various combinations of available security groups, then executing the ansible playbooks for each user in the test setup and even ssh-ing to the box and test what each user can do and can not do.
To improve validation and streamline future changes, this PR should include an automated test suite that verifies the following:
	•	Users are provisioned according to the specifications defined in `users.csv`.
	•	Users who are allowed to run playbooks can do so successfully.
	•	Users who are not permitted to run playbooks are correctly restricted.
Adding these tests will not only help validate this PR but also serve as a foundation for safely evolving the security setup over time, ensuring that all defined security rules continue to be enforced.

daryl375 commented on ansible/roles/iam_manager/templates/reset-my-password.j2:
```diff
@@ -0,0 +1,392 @@
+    for ((i=1; i<iterations; i++)); do
+        hash=$(printf %s "$hash" | sha256sum | awk '{print $1}')
+    done
+    echo "$hash"
```
Replaced custom SHA256 iteration hashing with system utilities (openssl or mkpasswd) for secure password hashing in reset-my-password.j2.

daryl375 commented on ansible/roles/iam_manager/templates/reset-my-password.j2:
```diff
@@ -0,0 +1,392 @@
+        if [[ "$input" == *"${dangerous_chars:$i:1}"* ]]; then
+            return 1
+        fi
+    done
```
Updated input validation to use a whitelist of safe characters and improved error messages.

daryl375 commented on ansible/roles/iam_manager/templates/reset-my-password.j2:
```diff
@@ -0,0 +1,388 @@
+            local xor_code=$((char_code ^ key_code))
+            result+=$(printf '\\x%02x' $xor_code)
+        done
+        printf '%s' "$result" | base64 -w 0
```
Adds error handling to the encrypt_data function in the password reset template to gracefully fail when no strong encryption tool (OpenSSL) is available, improving security and user feedback.

daryl375 commented on ansible/roles/iam_manager/tasks/create_users.yml:
```diff
@@ -1,56 +1,78 @@
-    groups: "{{ (item.value.group_a | default('') | lower + ',' + item.value.group_b | default('') | lower) | regex_replace('^,|,$', '') | regex_replace(',,', ',') | regex_replace('^$', '') }}"
+    shell: /bin/bash
+    state: present
+    groups: "{{ (item.value.group_a | default('') + ',' + item.value.group_b | default('') + ',' + item.value.group_c | default('')) | regex_replace('^,|,$', '') | regex_replace(',,', ',') | regex_replace('^$', '') }}"
```
Refactored group assignment in create_users.yml to use a more robust and readable approach for joining user groups, ensuring empty and undefined values are excluded.

daryl375 commented:
Introduces a README and CI/CD-friendly shell scripts for running Ansible Molecule test scenarios in the Hayek Validator Kit. Scripts support running all scenarios, individual scenarios, and provide features like parameter handling, and detailed reporting. The README documents usage, available scenarios, and integration tips.
Changed all references from /root/new-metal-box and ~/new-metal-box to /new-metal-box in Docker Compose, Ansible tasks, and setup scripts for consistency and container compatibility. Added generate-test-csv.sh script to automate test user SSH key and CSV generation at container startup.
Reviewed test tasks; test isolation not required since no cross-affecting changes

daryl375 commented on ansible-tests/molecule/iam_manager_tests/tasks/logging.yml:
```diff
@@ -0,0 +1,238 @@
+- name: Save detailed execution log to file
+  ansible.builtin.copy:
+    content: "{{ detailed_execution_log | join('\n') }}"
+    dest: "/test-reports/rbac_execution_log_{{ ansible_date_time.epoch }}.txt"
```
Introduces a set_fact task to define the reports_dir variable, ensuring consistent use of the report directory path. Updates the log file destination to use this variable in the copy task.

daryl375 commented on ansible/roles/iam_manager/templates/reset-my-password.j2:
```diff
@@ -0,0 +1,411 @@
+        else
+            return 1
+        fi
+        fi
```
The encrypt_data function now attempts AES-256-GCM encryption first, falling back to AES-256-CBC only if GCM fails. This improves security and simplifies the control flow. Also fixed a minor formatting issue in the XOR fallback.

daryl375 commented on ansible/roles/iam_manager/templates/reset-my-password.j2:
```diff
@@ -0,0 +1,398 @@
+}
+
+# Decrypt data using AES-256-GCM
+decrypt_data() {
+    local encrypted_data="$1"
+    local key="$2"
+
+    if command -v openssl >/dev/null 2>&1; then
+        # Split iv:encrypted_data
+        local iv="${encrypted_data%%:*}"
+        local encrypted="${encrypted_data#*:}"
+
+        # Decrypt with AES-256-CBC
+        local decrypted=$(printf '%s' "$encrypted" | openssl enc -aes-256-cbc -d -a -A -K "$key" -iv "$iv" 2>/dev/null)
```
The decrypt_data function now detects and handles AES-256-GCM encrypted data in addition to the existing AES-256-CBC format. This improves compatibility with data encrypted using GCM mode by parsing the IV, ciphertext, and tag, and using the appropriate openssl command.

daryl375 commented on ansible/roles/iam_manager/templates/30-validator-operators.j2:
```diff
@@ -0,0 +1,75 @@
+%validator_operators ALL=(ALL) NOPASSWD: /usr/bin/apt-get upgrade
+%validator_operators ALL=(ALL) NOPASSWD: /usr/bin/apt upgrade
+%validator_operators ALL=(ALL) NOPASSWD: /usr/bin/apt-get remove
+%validator_operators ALL=(ALL) NOPASSWD: /usr/bin/apt remove
```
Replaces wildcard apt install/remove permissions for validator_operators with explicit package lists in the sudoers template. This change improves security by limiting which packages can be installed or removed via sudo, reducing the risk of unauthorized system modifications.

daryl375 commented on ansible/roles/iam_manager/templates/reset-my-password.j2:
```diff
@@ -0,0 +1,398 @@
+}
+
+# Enhanced input validation with better sanitization
+validate_input() {
+    local input="$1"
+    local min_length="${2:-6}"
+
+    if [[ ${#input} -lt $min_length ]]; then
+        return 1
+    fi
+
+    # Whitelist: allow alphanumerics, spaces, and safe punctuation
+    # Safe characters: letters, numbers, space, period, comma, question mark, exclamation, hyphen, underscore
+    local allowed_pattern='^[a-zA-Z0-9 .,?!_][a-zA-Z0-9 .,?!_-]*$'
+    if ! [[ "$input" =~ $allowed_pattern ]]; then
+        echo "Invalid characters detected. Only letters, numbers, spaces, and basic punctuation (.,?!_-) are allowed."
+        return 1
+    fi
```
The input validation pattern now permits additional symbols (@, #, $, /, %, &) for password reset. The error message has been updated to reflect the expanded set of allowed characters.

daryl375 commented on ansible/roles/iam_manager/tasks/create_users.yml:
```diff
@@ -61,20 +83,110 @@
+
+- name: Extract SSH key part
+  ansible.builtin.set_fact:
+    ssh_keys: "{{ ssh_keys | default({}) | combine({item.key: (item.value.key.split(' ')[:2] | join(' '))}) }}"
```
Updated the 'Extract SSH key part' task to store the entire SSH key instead of just the first two space-separated parts. This ensures the full key is available for subsequent operations.

daryl375 commented on ansible/roles/iam_manager/templates/30-validator-operators.j2:
```diff
@@ -0,0 +1,199 @@
+%validator_operators ALL=(ALL) NOPASSWD: /usr/bin/strace -p
+
+# Shell access (requires password)
+%validator_operators ALL=(ALL) PASSWD: /bin/sh
```
to do

daryl375 commented on ansible/roles/iam_manager/templates/30-validator-operators.j2:
```diff
@@ -0,0 +1,199 @@
+%validator_operators ALL=(ALL) NOPASSWD: /usr/bin/systemctl disable {{ validator_service }}
+
+# Package management commands
+%validator_operators ALL=(ALL) NOPASSWD: /usr/bin/apt-get install apt-utils
+%validator_operators ALL=(ALL) NOPASSWD: /usr/bin/apt-get install locales
+%validator_operators ALL=(ALL) NOPASSWD: /usr/bin/apt-get install python3-dev
+%validator_operators ALL=(ALL) NOPASSWD: /usr/bin/apt-get install python3-setuptools
+%validator_operators ALL=(ALL) NOPASSWD: /usr/bin/apt-get install python3-yaml
+%validator_operators ALL=(ALL) NOPASSWD: /usr/bin/apt-get install software-properties-common
+%validator_operators ALL=(ALL) NOPASSWD: /usr/bin/apt-get install rsyslog
+%validator_operators ALL=(ALL) NOPASSWD: /usr/bin/apt-get install sudo
+%validator_operators ALL=(ALL) NOPASSWD: /usr/bin/apt-get install iproute2
+%validator_operators ALL=(ALL) NOPASSWD: /usr/bin/apt-get install openssh-client
+%validator_operators ALL=(ALL) NOPASSWD: /usr/bin/apt-get install openssh-server
+%validator_operators ALL=(ALL) NOPASSWD: /usr/bin/apt-get install nano
+%validator_operators ALL=(ALL) NOPASSWD: /usr/bin/apt-get install openssl
+%validator_operators ALL=(ALL) NOPASSWD: /usr/bin/apt-get install tar
+%validator_operators ALL=(ALL) NOPASSWD: /usr/bin/apt-get install less
+%validator_operators ALL=(ALL) NOPASSWD: /usr/bin/apt-get install tree
+%validator_operators ALL=(ALL) NOPASSWD: /usr/bin/apt-get install at
+%validator_operators ALL=(ALL) NOPASSWD: /usr/bin/apt-get install htop
+%validator_operators ALL=(ALL) NOPASSWD: /usr/bin/apt install apt-utils
+%validator_operators ALL=(ALL) NOPASSWD: /usr/bin/apt install locales
+%validator_operators ALL=(ALL) NOPASSWD: /usr/bin/apt install python3-dev
+%validator_operators ALL=(ALL) NOPASSWD: /usr/bin/apt install python3-setuptools
+%validator_operators ALL=(ALL) NOPASSWD: /usr/bin/apt install python3-yaml
+%validator_operators ALL=(ALL) NOPASSWD: /usr/bin/apt install software-properties-common
+%validator_operators ALL=(ALL) NOPASSWD: /usr/bin/apt install rsyslog
+%validator_operators ALL=(ALL) NOPASSWD: /usr/bin/apt install sudo
```
these packages are necessary for install jito

daryl375 commented on ansible/roles/iam_manager/templates/reset-my-password.j2:
```diff
@@ -0,0 +1,409 @@
+    # Safe characters: letters, numbers, space, period, comma, question mark, exclamation, hyphen, underscore
+    local allowed_pattern='^[a-zA-Z0-9 .,?!_@#$/%&-][a-zA-Z0-9 .,?!_@#$/%&-]*$'
+    if ! [[ "$input" =~ $allowed_pattern ]]; then
+    echo "Invalid characters detected. Only letters, numbers, spaces, and these symbols are allowed: . , ? ! _ - @ # $ / % &"
```
Modified the regex pattern and error message in the password reset template to clarify and adjust the set of allowed special characters for input validation.

daryl375 commented on ansible/roles/iam_manager/templates/reset-my-password.j2:
```diff
@@ -0,0 +1,409 @@
+        # Check if mkpasswd supports argon2
+        if mkpasswd --help 2>&1 | grep -q 'argon2'; then
+            # Use mkpasswd with argon2 if available
+            mkpasswd --method=argon2 --salt="$salt" "$password"
```
Added specific memory cost, time cost, and parallelism options to Argon2 hashing in the reset-my-password.j2 template for improved security and consistency.

daryl375 commented on ansible-tests/molecule/iam_manager_tests/tasks/rbac_tests/execution_tests.yml:
```diff
@@ -0,0 +1,187 @@
+      {{ tests | to_json }}
+
+- name: Execute command matrix with sudo -n for privilege testing
+  ansible.builtin.shell: "su - {{ item.user }} -c '{{ item.cmd | replace('{{ validator_service }}', validator_service) | replace('{{ validator_log_file }}', validator_log_file) | replace('{{ solana_user }}', solana_user) }}'"
```
Replaces the use of the ansible.builtin.shell module with ansible.builtin.command for executing privilege test commands. This change improves security and reliability by avoiding shell invocation and using direct command execution.

daryl375 commented on ansible/roles/iam_manager/tasks/disable_ubuntu.yml:
disable ubuntu user completely in a single task

- Refactored to consolidate all steps (lock account, remove sudo/admin groups, clean sudoers, deny SSH) into one task.
- Improved operator message to explicitly warn that 'ubuntu' user will be disabled and SSH sessions must use alice, bob, or carla instead.
- Updated documentation to fix incorrect inventory file references.
- Provisioned a bare metal server in Latitude (c2-small) for testing these changes.
